doc/topics/cloud/vagrant.rst

```diff
@@ -0,0 +1,304 @@
+.. _getting-started-with-vagrant:
+
+============================
+Getting Started With Vagrant
+============================
+
+The Vagrant driver is a new, experimental driver for spinning up a VagrantBox
+virtual machine, and installing Salt on it.
+
+Dependencies
+============
+The Vagrant driver itself has no external dependencies.
+
+The machine which will host the VagrantBox must be an already existing minion
+of the cloud server's Salt master.
+It must have Vagrant_ installed, and a Vagrant-compatible virtual machine engine,
+such as VirtualBox_.
+(Note: The Vagrant driver does not depend on the salt-cloud VirtualBox driver in any way.)
+
+.. _Vagrant: https://www.vagrantup.com/
+.. _VirtualBox: https://www.virtualbox.org/
+
+\[Caution: The version of Vagrant packaged for ``apt install`` in Ubuntu 16.04 will not connect a bridged
+network adapter correctly. Use a version downloaded directly from the web site.\]
+
+Include the Vagrant guest editions plugin:
+``vagrant plugin install vagrant-vbguest``.
+
+Salt-api must be installed and configured on the salt master.
```

Comment:
Why are we using the salt-api here?

I understand why it is being used for saltify, you have no other records of the information and have to ask salt for them.

IMO, we should just use vagrant directly and not salt-api. to get and gather information about the minions.

Reply:
I use salt calls to send the vagrant commands to the VM host. Vagrant does not have a network API, so commands have to be done using CLI commands on the host computer. I figured it was more of a sure thing to use salt "cmd.run" rather than ssh to do that. 

The only ssh command I use -- the one to get the VM's IP addresses -- was a pain. (See lines 248 to 298 of vagrant.py). I used "cmd.run" on the host to run a local copy of ssh.  Ugly.  Real ssh support would have required duplicating or calling the extensive ssh routines from salt/utils/cloud.py,  then setting up an ssh router on the host.

Perhaps it would be possible to use salt-ssh rather than running a minion on the VM host. I did not really investigate that option. That would still require salt-api, or something like it. I am trying to allow for the cloud master to be on a different machine from the salt master. (But I have not tested that.)

Running the VMs on the salt-master machine is easy. Perhaps that is why it is a restriction for the salt-cloud VirtualBox driver. Running the VMs on a different machine -- that was hard even with Salt to do the grunt work.

Reply:
Salt-cloud is supposed to reach out to the utility that is running vagrant, and run commands.

This is a limitation of the VirtualBox stuff, but libvirt can be used to talk to VirtualBox over the network and then the [libvirt driver](https://docs.saltstack.com/en/latest/topics/cloud/libvirt.html) can be used.

https://libvirt.org/drvvbox.html

I am not going to approve this as is, it needs to use the vagrant command directly and only build on the salt-master, it will have the same limitations as using the virtualbox driver correctly, but I do not like adding this.

Reply:
With this driver, salt-cloud _does_ reach out to the utility that is running Vagrant -- it uses salt-api as the tool to do it. No other tool works as well. 

Running the VM only on the salt-master is next to useless. 
[Refer to xkcd](http://xkcd.com/1764)

Reply:
If I re-wrote this using Python calls to the API, would it be more acceptable?

Reply:
If you want to run vagrant as a hypervisor on other hosts, you should write something similar to SaltVirt.

https://docs.saltstack.com/en/latest/topics/virt/index.html

Reply:
By great good fortune, I am at the beginning of a two-week research sprint. I will check out how SaltVirt uses Salt to do what it needs on the host machine. Perhaps I can replace my salt-api calls with another communication method.

Reply:
Preliminary report:

It seems that I have been trying to  re-invent a wheel. The note in https://docs.saltstack.com/en/latest/topics/virt/index.html says that "Salt was originally developed with the intent of using the Salt communication system as the backbone to a cloud controller."

Nevertheless, SaltVirt does not use Saltstack communication to reach a remote hypervisor. It passes the remote URL to the libvirt package and makes it do all the work. It also uses routines in salt/modules/virt.py to perform many of its functions. There are some ideas there, especially how one might pass additional information (using environment variables and pillar values) to control the bootstrapped machine.  This could solve my dilemma of how to control a machine that I cannot get a ssh link to -- which is the cause of much of the complexity in my "create" function, and my biggest limitation in operation.

I don't think libvirt is going to work. Vagrant is not among the thirteen providers supported by it, and for good reason: vagrant is too high level. The libvirt xml files contain basically the same information contained in the Vagrantfile. In order to emulate libvirt commands using vagrant, libvirt would have to edit the Vagrantfile. The conversion from xml to ruby would be nasty -- especially since the ruby code is not very well documented. There be dragons. 

I am getting some ideas from clouds/libvirt.py about how they handle the ssh connection for salt-bootstrap. It makes a call to salt.utils.cloud-wait_for_ip and I was afraid to try connection at that level.